packages/utils/src/abstract-session.ts

```diff
@@ -0,0 +1,221 @@
+    super()
+
+    setMaxListeners(Infinity, this)
+    this.name = init.name
+    this.logger = components.logger
+    this.log = components.logger.forComponent(this.name)
+    this.requests = new Map()
+    this.minProviders = init.minProviders ?? DEFAULT_SESSION_MIN_PROVIDERS
+    this.maxProviders = init.maxProviders ?? DEFAULT_SESSION_MAX_PROVIDERS
+    this.providers = []
+  }
+
+  async retrieve (cid: CID, options: BlockRetrievalOptions<RetrieveBlockProgressEvents> = {}): Promise<Uint8Array> {
+    // see if we are already requesting this CID in this session
+    const cidStr = base64.encode(cid.multihash.bytes)
+    const existingJob = this.requests.get(cidStr)
+
+    if (existingJob != null) {
+      this.log('join existing request for %c', cid)
+      return existingJob
+    }
+
+    const deferred: DeferredPromise<Uint8Array> = pDefer()
+    this.requests.set(cidStr, deferred.promise)
+
+    if (this.providers.length === 0) {
+      let first = false
+
+      if (this.intialPeerSearchComplete == null) {
+        first = true
+        this.log = this.logger.forComponent(`${this.name}:${cid}`)
+        this.intialPeerSearchComplete = this.findNewProviders(cid, this.minProviders, options)
+      }
+
+      await this.intialPeerSearchComplete
+
+      if (first) {
+        this.log('found initial session peers for %c', cid)
+      }
+    }
+
+    let foundBlock = false
+
+    // this queue manages outgoing requests - as new peers are added to the
+    // session they will be added to the queue so we can request the current
+    // block from multiple peers as they are discovered
+    const queue = new Queue<Uint8Array, { provider: Provider, priority?: number }>({
+      concurrency: this.maxProviders
+    })
+    queue.addEventListener('error', (evt) => {
+      if (options.signal?.aborted === true) {
+        deferred.reject(new AbortError('Block want was aborted'))
+      }
+    })
+    queue.addEventListener('failure', (evt) => {
+      if (options.signal?.aborted === true) {
+        deferred.reject(new AbortError('Block want was aborted'))
+        return
+      }
+
+      this.log.error('error querying provider %o', evt.detail.job.options.provider, evt.detail.error)
+    })
+    queue.addEventListener('success', (evt) => {
+      if (options.signal?.aborted === true) {
+        deferred.reject(new AbortError('Block want was aborted'))
+        return
+      }
+
+      // peer has sent block, return it to the caller
+      foundBlock = true
+      deferred.resolve(evt.detail.result)
+    })
+    queue.addEventListener('idle', () => {
+      if (options.signal?.aborted === true) {
+        deferred.reject(new AbortError('Block want was aborted'))
+        return
+      }
+
+      // we found it, all good
+      if (foundBlock) {
+        return
+      }
+
+      // find more session peers and retry
+      Promise.resolve()
+        .then(async () => {
```

Comment:
same as elsewhere, why promise.resolve.then().catch() instead of try/catch?

also, linter should be catching this and required a prefixed `void`
```suggestion
      void Promise.resolve()
        .then(async () => {
```

Reply:
Answered elsewhere: https://github.com/ipfs/helia/pull/495#discussion_r1563942408

> linter should be catching this and required a prefixed void

`void` is only required if it's a floating promise.  There's a `.catch(...)` here so it's not floating, so no `void` required by the linter.